providers/implementations/signature/ml_dsa_sig.c

```diff
@@ -0,0 +1,245 @@
+{
+    return ml_dsa_signverify_msg_init(vctx, vkey, params,
+                                      EVP_PKEY_OP_SIGN, "ML_DSA Sign Init");
+}
+
+static int ml_dsa_sign(void *vctx, unsigned char *sig, size_t *siglen,
+                       size_t sigsize, const unsigned char *msg, size_t msg_len)
+{
+    int ret = 0;
+    PROV_ML_DSA_CTX *ctx = (PROV_ML_DSA_CTX *)vctx;
+    uint8_t rand_tmp[ML_DSA_ENTROPY_LEN], *rnd = NULL;
+
+    if (!ossl_prov_is_running())
+        return 0;
+
+    if (sig != NULL) {
+        if (ctx->test_entropy_len != 0) {
+            rnd = ctx->test_entropy;
+        } else {
+            rnd = rand_tmp;
+
+            if (ctx->deterministic == 1)
+                memset(rnd, 0, sizeof(rand_tmp));
+            else if (RAND_priv_bytes_ex(ctx->libctx, rnd, sizeof(rand_tmp), 0) <= 0)
+                return 0;
+        }
+    }
+    ret = ossl_ml_dsa_sign(ctx->ctx, ctx->key, msg, msg_len,
+                           ctx->context_string, ctx->context_string_len,
+                           rnd, sizeof(rand_tmp), ctx->msg_encode,
+                           sig, siglen, sigsize);
+    if (rnd != ctx->test_entropy)
+        OPENSSL_cleanse(rand_tmp, sizeof(rand_tmp));
+    return ret;
+}
+
+static int ml_dsa_verify_msg_init(void *vctx, void *vkey, const OSSL_PARAM params[])
+{
+    return ml_dsa_signverify_msg_init(vctx, vkey, params, EVP_PKEY_OP_VERIFY,
+                                   "ML_DSA Verify Init");
+}
+
+static int ml_dsa_verify(void *vctx, const unsigned char *sig, size_t siglen,
+                         const unsigned char *msg, size_t msg_len)
+{
+    PROV_ML_DSA_CTX *ctx = (PROV_ML_DSA_CTX *)vctx;
+
+    if (!ossl_prov_is_running())
+        return 0;
+    return ossl_ml_dsa_verify(ctx->ctx, ctx->key, msg, msg_len,
+                              ctx->context_string, ctx->context_string_len,
+                              ctx->msg_encode, sig, siglen);
+}
+
+static int ml_dsa_set_ctx_params(void *vctx, const OSSL_PARAM params[])
+{
+    PROV_ML_DSA_CTX *pctx = (PROV_ML_DSA_CTX *)vctx;
+    const OSSL_PARAM *p;
+
+    if (pctx == NULL)
+        return 0;
+    if (ossl_param_is_empty(params))
+        return 1;
+
+    p = OSSL_PARAM_locate_const(params, OSSL_SIGNATURE_PARAM_CONTEXT_STRING);
+    if (p != NULL) {
+        void *vp = pctx->context_string;
+
+        if (!OSSL_PARAM_get_octet_string(p, &vp, sizeof(pctx->context_string),
+                                         &(pctx->context_string_len))) {
+            pctx->context_string_len = 0;
+            return 0;
+        }
+    }
+    p = OSSL_PARAM_locate_const(params, OSSL_SIGNATURE_PARAM_TEST_ENTROPY);
+    if (p != NULL) {
+        void *vp = pctx->test_entropy;
+
+        if (!OSSL_PARAM_get_octet_string(p, &vp, sizeof(pctx->test_entropy),
+                                         &(pctx->test_entropy_len))
+                || pctx->test_entropy_len != sizeof(pctx->test_entropy)) {
```

Comment:
The length mismatch definitely needs an ERR_raise_data to indicate the issue ...